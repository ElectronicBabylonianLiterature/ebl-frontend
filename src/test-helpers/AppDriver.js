import React from 'react'
import { fireEvent, render, waitForElement } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import App from 'App'
import Auth from 'auth/Auth'
import WordRepository from 'dictionary/WordRepository'
import FragmentRepository from 'fragmentarium/FragmentRepository'
import ImageRepository from 'fragmentarium/ImageRepository'
import FragmentService from 'fragmentarium/FragmentService'
import WordService from 'dictionary/WordService'
import TextService from 'corpus/TextService'
import Session from 'auth/Session'
import BibliographyRepository from 'bibliography/BibliographyRepository'
import BibliographyService from 'bibliography/BibliographyService'
import { defaultErrorReporter } from 'ErrorReporterContext'
import createAuth0Config from 'auth/createAuth0Config'

function createApp(api, sessionStore) {
  const auth0Config = createAuth0Config()
  const auth = new Auth(sessionStore, defaultErrorReporter, auth0Config)
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    bibliographyService
  )
  const textService = new TextService(api)

  return (
    <App
      auth={auth}
      wordService={wordService}
      fragmentService={fragmentService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  initialEntries = []
  element = null
  session = null

  constructor(api) {
    this.api = api
  }

  withPath(path) {
    this.initialEntries = [path]
    return this
  }

  withSession() {
    this.session = new Session(
      'accessToken',
      'idToken',
      Number.MAX_SAFE_INTEGER,
      ['write:texts']
    )
    return this
  }

  render() {
    const fakeSessionStore = {
      setSession: session => {
        this.session = session
      },
      clearSession: () => {
        this.session = null
      },
      getSession: () => this.session || new Session('', '', 0, [])
    }

    this.element = render(
      <MemoryRouter initialEntries={this.initialEntries}>
        {createApp(this.api, fakeSessionStore)}
      </MemoryRouter>
    )

    return this
  }

  async waitForText(text) {
    await waitForElement(() => this.element.getByText(text))
  }

  expectTextContent(text) {
    expect(this.element.container).toHaveTextContent(text)
  }

  expectNotInContent(text) {
    expect(this.element.queryByText(text)).toBeNull()
  }

  expectLink(text, expectedHref) {
    expect(this.element.getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs) {
    this.expectTextContent(crumbs.join(''))
  }

  expectInputElement(label, expectedValue) {
    expect(this.element.getByLabelText(label).value).toEqual(
      String(expectedValue)
    )
  }

  changeValueByLabel(label, newValue) {
    const input = this.element.getByLabelText(label)
    fireEvent.change(input, { target: { value: newValue } })
  }

  click(text, n = 0) {
    const clickable = this.element.getAllByText(text)[n]
    fireEvent.click(clickable)
    // await wait()
  }
}
