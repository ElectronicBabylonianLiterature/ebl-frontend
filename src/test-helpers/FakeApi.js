import Promise from 'bluebird'

export default class FakeApi {
  gets = []
  client = {
    fetchJson: (path, authenticate) => {
      const expectation = this.gets.find(entry => entry.path === path && entry.authenticate === authenticate)
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchJson: ${path}`))
    },

    fetchBlob: (path, authenticate) => {
      return Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchBlob: ${path}`))
    }
  }

  expectText (text) {
    this.gets.push({
      path: `/texts/${text.category}/${text.index}`,
      authenticate: true,
      response: text
    })
    return this
  }

  expectStatistics (statistics) {
    this.gets.push({
      path: `/statistics`,
      authenticate: false,
      response: statistics
    })
    return this
  }
}
