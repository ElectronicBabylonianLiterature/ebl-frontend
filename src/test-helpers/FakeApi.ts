import Promise from 'bluebird'
import Word from 'dictionary/domain/Word'

class Expectation {
  method: 'POST' | 'GET' = 'GET'
  path = ''
  authenticate = true
  response = {}
  verify = false
  called = false
  body: object | null = null
  isBlob = false

  constructor(data) {
    Object.assign(this, data)
  }
}

export default class FakeApi {
  private expectations: Expectation[] = []

  readonly client = {
    fetchJson: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchJson: ${path}`
            )
          )
    }),

    postJson: jest.fn().mockImplementation(path => {
      const expectation = this.expectations.find(
        entry => entry.method === 'POST' && entry.path === path && !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected postJson: ${path}`))
    }),

    fetchBlob: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchBlob: ${path}`
            )
          )
    })
  }

  expectTexts(texts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts`,
        authenticate: false,
        response: texts,
        verify: true
      })
    )
    return this
  }

  allowText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text
      })
    )
    return this
  }

  expectText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text,
        verify: true
      })
    )
    return this
  }

  expectUpdateManuscripts(text, chapterIndex, manuscripts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/manuscripts`,
        response: text,
        verify: true,
        body: manuscripts
      })
    )
    return this
  }

  expectUpdateLines(text, chapterIndex, lines): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/lines`,
        response: text,
        verify: true,
        body: lines
      })
    )
    return this
  }

  expectAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/annotations`,
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectUpdateAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/fragments/${number}/annotations`,
        body: {
          fragmentNumber: number,
          annotations: annotationDtos
        },
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectFragment(fragmentDto): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${fragmentDto._id}`,
        authenticate: true,
        response: fragmentDto,
        verify: true
      })
    )
    return this
  }

  expectPhoto(number: string): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/photo`,
        authenticate: true,
        isBlob: true,
        verify: true
      })
    )
    return this
  }

  expectSearchWords(query: string, words: readonly Word[]): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words?query=${encodeURIComponent(query)}`,
        response: words,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words/${encodeURIComponent(word._id)}`,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectUpdateWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/words/${encodeURIComponent(word._id)}`,
        body: word,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  allowStatistics(statistics): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/statistics`,
        authenticate: false,
        response: statistics
      })
    )
    return this
  }

  allowImage(file): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/images/${file}`,
        authenticate: false,
        isBlob: true
      })
    )
    return this
  }

  verifyExpectations(): void {
    const methods = {
      GET: (expectation: Expectation): void =>
        expect(
          expectation.isBlob ? this.client.fetchBlob : this.client.fetchJson
        ).toHaveBeenCalledWith(expectation.path, expectation.authenticate),
      POST: (expectation: Expectation): void =>
        expect(this.client.postJson).toHaveBeenCalledWith(
          expectation.path,
          expectation.body || expect.anything()
        )
    }
    this.expectations
      .filter(expectation => expectation.verify)
      .forEach(expectation => methods[expectation.method](expectation))
  }
}
