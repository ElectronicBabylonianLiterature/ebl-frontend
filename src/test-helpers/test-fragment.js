import {
  Folio,
  Fragment,
  Line,
  Measures,
  RecordEntry,
  Text
} from '../fragmentarium/fragment'
import { List, Map } from 'immutable'

export const fragmentDto = {
  _id: 'Test.Fragment',
  accession: '',
  cdliNumber: '',
  bmIdNumber: '',
  publication: 'Electronic Babylonian Literature',
  description: 'A fragment to be used when testing the eBL application',
  joins: [],
  length: { value: 3.14, note: '(complete)' },
  width: { value: 0.30282212, note: '' },
  thickness: {},
  collection: '',
  script: 'NB',
  notes: '',
  museum: '',
  signs: 'SAL/P₂ ŠIM GU GA\nP₅/SAL ŠIM\nŠIM ŠIM\nKU KA GA KU X',
  record: [
    { user: 'Laasonen', type: 'Revision', date: '2019-02-01T14:22:40.201231' },
    { user: 'Laasonen', type: 'Revision', date: '2019-02-01T14:23:25.214166' }
  ],
  folios: [{ name: 'FWG', number: 'M134' }],
  text: {
    lines: [
      {
        prefix: '10.',
        content: [
          {
            type: 'Word',
            value: 'sal/:',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'Word',
            value: 'š[im',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          },
          {
            type: 'LoneDeterminative',
            value: '{gu}',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: [false, true]
          },
          { type: 'Token', value: '[...]' },
          {
            type: 'Word',
            value: '.GA',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }
        ],
        type: 'TextLine'
      },
      {
        prefix: '10.',
        content: [
          {
            type: 'Word',
            value: '::/sal',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'Word',
            value: 'ši]m',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }
        ],
        type: 'TextLine'
      },
      {
        prefix: '10.',
        content: [
          {
            type: 'Word',
            value: 'šim',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          },
          { type: 'Token', value: '|' },
          {
            type: 'Word',
            value: 'šim',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }
        ],
        type: 'TextLine'
      },
      {
        prefix: '10.',
        content: [
          { type: 'Token', value: '...' },
          {
            type: 'Word',
            value: '+ku',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'LoneDeterminative',
            value: '{KA.G[A}',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: [false, false]
          },
          { type: 'Token', value: '...' },
          { type: 'Token', value: '...' },
          {
            type: 'Word',
            value: '.ku',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'Word',
            value: 'x',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }
        ],
        type: 'TextLine'
      }
    ]
  },
  references: [
    { id: 'RN52', type: 'DISCUSSION', pages: '', notes: '', linesCited: [] }
  ],
  atf:
    '10. sal/: š[im {gu}[...].GA\n10. ::/sal ši]m\n10. šim | šim\n10. ...+ku {KA.G[A} ... ....ku x'
}

export const fragment = new Fragment({
  number: 'Test.Fragment',
  accession: '',
  cdliNumber: '',
  bmIdNumber: '',
  publication: 'Electronic Babylonian Literature',
  description: 'A fragment to be used when testing the eBL application',
  joins: List(),
  measures: Measures({
    length: 3.14,
    width: 0.30282212,
    thickness: null
  }),
  collection: '',
  script: 'NB',
  notes: '',
  museum: '',
  record: List.of(
    new RecordEntry({
      user: 'Laasonen',
      date: '2019-02-01T14:22:40.201231',
      type: 'Revision'
    }),
    new RecordEntry({
      user: 'Laasonen',
      date: '2019-02-01T14:23:25.214166',
      type: 'Revision'
    })
  ),
  folios: List([new Folio({ name: 'FWG', number: 'M134' })]),
  text: Text({
    lines: List.of(
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({
            type: 'Word',
            value: 'sal/:',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'Word',
            value: 'š[im',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }),
          Map({
            type: 'LoneDeterminative',
            value: '{gu}',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: List([false, true])
          }),
          Map({ type: 'Token', value: '[...]' }),
          Map({
            type: 'Word',
            value: '.GA',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          })
        )
      }),
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({
            type: 'Word',
            value: '::/sal',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'Word',
            value: 'ši]m',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          })
        )
      }),
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({
            type: 'Word',
            value: 'šim',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }),
          Map({ type: 'Token', value: '|' }),
          Map({
            type: 'Word',
            value: 'šim',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          })
        )
      }),
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({ type: 'Token', value: '...' }),
          Map({
            type: 'Word',
            value: '+ku',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'LoneDeterminative',
            value: '{KA.G[A}',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: List([false, false])
          }),
          Map({ type: 'Token', value: '...' }),
          Map({ type: 'Token', value: '...' }),
          Map({
            type: 'Word',
            value: '.ku',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'Word',
            value: 'x',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          })
        )
      })
    )
  }),
  references: List([
    Map({
      id: 'RN52',
      type: 'DISCUSSION',
      pages: '',
      notes: '',
      linesCited: List()
    })
  ]),
  atf:
    '10. sal/: š[im {gu}[...].GA\n10. ::/sal ši]m\n10. šim | šim\n10. ...+ku {KA.G[A} ... ....ku x'
})
