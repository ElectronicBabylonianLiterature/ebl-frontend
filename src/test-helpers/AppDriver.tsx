import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-helpers/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api)

  return (
    <App
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private element: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getElement(): RenderResult {
    if (this.element) {
      return this.element
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words'
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.element = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              }
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getElement().findByText(text)
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getElement().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getElement().queryByText(text)).toBeNull()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getElement().getByText(text)).toHaveAttribute(
      'href',
      expectedHref
    )
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getElement().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(
      (this.getElement().getByLabelText(label) as HTMLInputElement).value
    ).toEqual(String(expectedValue))
  }

  changeValueByLabel(label: Matcher, newValue): void {
    const input = this.getElement().getByLabelText(label)
    act(() => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getElement())
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getElement().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
