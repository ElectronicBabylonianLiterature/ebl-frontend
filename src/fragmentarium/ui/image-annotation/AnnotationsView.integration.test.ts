import AppDriver from 'test-support/AppDriver'
import FakeApi from 'test-support/FakeApi'
import { fragmentDto } from 'test-support/test-fragment'
import { annotationsDto } from 'test-support/test-annotation'
import produce from 'immer'

const fragmentWithoutReferences = produce(fragmentDto, (draft) => {
  draft.references = []
})
const fragmentNumber = 'Test.Fragment'
const photo = { blobParts: [''], options: { type: 'image/jpeg' }, size: 1 }
let fakeApi: FakeApi
let appDriver: AppDriver

afterEach(() => {
  fakeApi.verifyExpectations()
})

describe('Display annotate view', () => {
  beforeEach(async () => {
    ;(URL.createObjectURL as jest.Mock).mockReturnValueOnce('mock url')
    fakeApi = new FakeApi()
      .expectFragment(fragmentWithoutReferences)
      .expectPhoto(fragmentNumber, photo)
      .expectAnnotations(fragmentNumber, annotationsDto)
    appDriver = new AppDriver(fakeApi.client)
      .withSession()
      .withPath(`/library/${fragmentNumber}/annotate`)
      .render()
    await appDriver.waitForText('blank')
  })

  test('Breadcrumbs', () => {
    appDriver.breadcrumbs.expectCrumbs([
      'eBL',
      'Fragmentarium',
      fragmentNumber,
      'Tag Signs',
    ])
  })

  test('Fragment crumb', () => {
    appDriver.breadcrumbs.expectCrumb(
      fragmentNumber,
      `/library/${fragmentNumber}`
    )
  })

  test('Snapshot', () => {
    expect(appDriver.getView().container).toMatchSnapshot()
  })
})
