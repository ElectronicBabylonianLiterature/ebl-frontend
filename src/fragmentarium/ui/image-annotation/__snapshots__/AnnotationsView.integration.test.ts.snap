// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display annotate view Snapshot 1`] = `
<div>
  <div
    class="main-body"
  >
    <header
      class="Header"
    >
      <nav
        class="navbar navbar-expand-md navbar-light"
      >
        <div
          class="container"
        >
          <a
            class="active navbar-brand"
            href="/"
            title="electronic Babylonian Library (eBL)"
          >
            <h1
              class="Header__title"
            >
              <span
                class="Header__title-main"
              >
                electronic
                <br />
                Babylonian
                <br />
                Library
              </span>
              <small
                class="Header__title-abbreviation"
              >
                eBL
              </small>
            </h1>
          </a>
          <span
            class="navbar-brand"
          >
            <div
              class="Header__logo-container container"
            >
              <a
                href="https://www.lmu.de"
                rel="noopener noreferrer"
                target="_blank"
              >
                <img
                  class="Header__lmu-logo img-fluid"
                  src="LMU_Logo.svg"
                />
              </a>
              <a
                href="https://badw.de/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <img
                  class="Header__badw-logo img-fluid"
                  src="BAdW_Logo.svg"
                />
              </a>
            </div>
          </span>
          <button
            aria-controls="Header-7"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
            id="Header-7"
          >
            <div
              class="d-flex justify-content-between"
              id="navbar-container"
            >
              <div
                id="menu-lines"
              >
                <div
                  class="mx-auto navbar-nav"
                >
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/signs"
                      href="/signs"
                    >
                      Signs
                    </a>
                  </div>
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/dictionary"
                      href="/dictionary"
                    >
                      Dictionary
                    </a>
                  </div>
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/corpus"
                      href="/corpus"
                    >
                      Corpus
                    </a>
                  </div>
                  <div
                    class="nav-item"
                  >
                    <a
                      class="active nav-link"
                      data-rb-event-key="/library"
                      href="/library"
                    >
                      Fragmentarium
                    </a>
                  </div>
                </div>
                <div
                  class="mx-auto navbar-nav"
                >
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/about"
                      href="/about"
                    >
                      About
                    </a>
                  </div>
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/bibliography"
                      href="/bibliography"
                    >
                      Bibliography
                    </a>
                  </div>
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/tools"
                      href="/tools"
                    >
                      Tools
                    </a>
                  </div>
                  <div
                    class="nav-item"
                  >
                    <a
                      class="nav-link"
                      data-rb-event-key="/projects"
                      href="/projects"
                    >
                      Projects
                    </a>
                  </div>
                </div>
              </div>
              <span
                class="navbar-text"
                id="user"
              >
                <button
                  class="btn btn-outline-secondary btn-sm"
                  type="button"
                >
                  Logout 
                  Test
                </button>
              </span>
            </div>
          </div>
        </div>
      </nav>
    </header>
    <main
      class="main main--wide"
    >
      <div
        class="main__content"
      >
        <header
          class="main__header"
        >
          <nav
            aria-label="breadcrumb"
            class="main__breadcrumbs"
          >
            <ol
              class="breadcrumb"
            >
              <li
                class="breadcrumb-item active"
              >
                <a
                  href="/"
                >
                  eBL
                </a>
              </li>
              <li
                class="breadcrumb-item active"
              >
                <a
                  href="/library"
                >
                  Fragmentarium
                </a>
              </li>
              <li
                class="breadcrumb-item active"
              >
                <a
                  href="/library/Test.Fragment"
                >
                  Test.Fragment
                </a>
              </li>
              <li
                aria-current="page"
                class="breadcrumb-item active"
              >
                Tag Signs
              </li>
            </ol>
          </nav>
          <h2
            class="main__heading"
          >
            Tag Signs
          </h2>
          <div
            class="main__toolbar btn-toolbar"
            role="toolbar"
          />
        </header>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-outline-dark"
            type="button"
          >
            Generate Annotations
          </button>
          <button
            class="btn btn-outline-dark"
            type="button"
          >
            Automatic Selection
          </button>
          <button
            class="btn btn-outline-dark"
            type="button"
          >
            Delete all
          </button>
          <button
            class="btn btn-outline-dark"
            type="button"
          >
            Save
          </button>
          <button
            class="btn btn-outline-dark"
            type="button"
          >
            Show Card
          </button>
        </div>
        <div
          class="ml-3  btn-group-vertical btn-group-sm"
          role="group"
        >
          <button
            class="btn btn-outline-dark"
            disabled=""
            type="button"
          >
            Mode: 
            default
          </button>
        </div>
        <i
          class="m-2 fas fa-info-circle"
        />
        <span
          class="text-danger"
        >
          ‚Üê Please read this!
        </span>
        <div
          class="row"
        >
          <div
            class="col-5"
          >
            <div
              class="mt-5 pt-4 row"
            >
              <div
                class="card"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-1"
                    >
                      <button
                        class="btn btn-outline-dark btn-sm"
                        disabled=""
                        type="button"
                      >
                        blank
                      </button>
                    </div>
                    <div
                      class="col-1"
                    >
                      <button
                        class="btn btn-outline-dark btn-sm"
                        disabled=""
                        type="button"
                      >
                        struct
                      </button>
                    </div>
                    <div
                      class="col-1"
                    >
                      <button
                        class="btn btn-outline-dark btn-sm"
                        disabled=""
                        type="button"
                      >
                        x
                      </button>
                    </div>
                    <div
                      class="col-4 offset-4"
                    />
                  </div>
                  <hr />
                  <div>
                    <span>
                      10.
                       
                    </span>
                    <span>
                      [...]
                       
                    </span>
                  </div>
                  <div>
                    <span>
                      10.
                       
                    </span>
                  </div>
                  <div>
                    <span>
                      10.
                       
                    </span>
                    <span>
                      |
                       
                    </span>
                  </div>
                  <div>
                    <span>
                      10.
                       
                    </span>
                    <span>
                      ...
                       
                    </span>
                    <span>
                      ...
                       
                    </span>
                    <span>
                      ...
                       
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mt-5 ml-0 mr-0 pl-1 pr-0 col-1"
          >
            <input
              class="h-75 ml-0 mr-0 pl-0 pr-0"
              orient="vertical"
              type="range"
              value="100"
            />
          </div>
          <div
            class="pl-0 col-6"
            style="padding-top: 0px; margin-left: -100px;"
          >
            <div
              class="my-3 row"
            >
              <div
                class="col-auto"
              >
                <div
                  class="btn-group"
                  role="group"
                >
                  <button
                    class="btn btn-outline-dark"
                    type="button"
                  >
                    Zoom In +
                  </button>
                  <button
                    class="btn btn-outline-dark"
                    type="button"
                  >
                    Zoom Out -
                  </button>
                  <button
                    class="btn btn-outline-dark"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
              </div>
              <div
                class="text-center my-auto col"
              >
                Zoom with Mouse wheel. Pan while holding shift
              </div>
            </div>
            <div
              class="react-transform-wrapper transform-component-module_wrapper__1_Fgj "
            >
              <div
                class="react-transform-component transform-component-module_content__2jYgh "
                style="transform: translate3d(0px, 0px, 0) scale(1);"
              >
                <div
                  class="sc-iwsKbI bNCCdn"
                >
                  <img
                    alt="Test.Fragment"
                    class="sc-gZMcBi gVmiLs"
                    draggable="false"
                    src="mock url"
                  />
                  <div
                    class="sc-gqjmRU inYJUn"
                  >
                    <div
                      data-testid="annotation__box"
                      style="transform: scale(0.3); transform-origin: top left; position: absolute; left: 100%; top: 45.7%; height: 13.333333333333334%; width: 0.06666666666666667%; border: 1px solid aliceblue; padding: 0px;"
                    >
                      <span
                        style="position: absolute; top: -1px; left: -1px; margin: 0px; padding: 0px; background: whitesmoke; opacity: 0.5; font-size: .8em;"
                      >
                        kur
                      </span>
                    </div>
                  </div>
                  <div
                    class="sc-gqjmRU inYJUn"
                    data-testid="annotation__target"
                  />
                  <div
                    class="sc-dnqmqq ixSjBM"
                  >
                    Click and Drag to Annotate
                  </div>
                  <div />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="main-footer"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/impressum"
          >
            Impressum
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/datenschutz"
          >
            Datenschutz
          </a>
        </li>
      </ol>
    </footer>
  </div>
</div>
`;
